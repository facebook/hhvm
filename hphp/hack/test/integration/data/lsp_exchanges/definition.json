[
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "initialize",
  "params":
    {
      "initializationOptions":{},
      "processId":null,
      "rootPath":"${root_path}",
      "capabilities":{}
    }
},
{
  "jsonrpc": "2.0",
  "method": "textDocument/didOpen",
  "params": {
     "textDocument": {
       "uri":"${php_file_uri}",
       "languageId": "hack",
       "version":1,
       "text":"${php_file}"
     }
   }
},
{
  "comment": "definition of `b_definition()`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 2,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 3, "character": 10}
   }
},
{
  "comment": "call to `new BB(1)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 3,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 29, "character": 13}
   }
},
{
  "comment": "call to `new CC(1)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 4,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 30, "character": 13}
   }
},
{
  "comment": "call to `new DD(1)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 5,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 31, "character": 13}
   }
},
{
  "comment": "call to `new EE(1)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 6,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 32, "character": 13}
   }
},
{
  "comment": "call to `new FF(1)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 7,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 33, "character": 13}
   }
},
{
  "comment": "call to `new TakesString(HasString::MyString)`",
  "jsonrpc": "2.0",
  "method": "textDocument/definition",
  "id": 8,
  "params": {
      "textDocument": {
        "uri":"${php_file_uri}"
       },
       "position": {"line": 45, "character": 23}
   }
},
{
  "jsonrpc": "2.0",
  "id": 99,
  "method": "shutdown",
  "params": {}
}
]
