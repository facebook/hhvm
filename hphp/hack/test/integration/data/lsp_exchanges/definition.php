<?hh  //strict

function a_definition(): int {
  return b_definition();
}

function b_definition(): int {
  return 42;
}

class BB {
  public function __construct(int $i) {}
}

class CC extends BB {
}

class DD extends CC {
}

class EE extends DD {
  public function __construct() {
    parent::__construct(1);
  }
}

class FF {}

function test(): void {
  $bb = new BB(1); // should go to B::__construct
  $cc = new CC(1); // should offer choice B::__construct or C
  $dd = new DD(1); // should offer choice B::__construct or D
  $ee = new EE(); // should go to E::__construct
  $ff = new FF(); // should go to F
}

class TakesString {
  public function __construct(string $s) {}
}

class HasString {
  const MyString = "myString";
}

function testClassMemberInsideConstructorInvocation(): void {
  $x = new TakesString(HasString::MyString);
}
