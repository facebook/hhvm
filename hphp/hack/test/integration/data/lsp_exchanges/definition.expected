[
  {
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
      "capabilities": {
        "textDocumentSync": {
          "openClose": true,
          "change": 2,
          "willSave": false,
          "willSaveWaitUntil": false,
          "save": {
            "includeText": false
          }
        },
        "hoverProvider": true,
        "completionProvider": {
          "resolveProvider": true,
          "triggerCharacters": [
            "$",
            ">",
            "\\",
            ":",
            "<",
            "[",
            "'",
            "\""
          ]
        },
        "signatureHelpProvider": {
          "triggerCharacters": [
            "(",
            ","
          ]
        },
        "definitionProvider": true,
        "typeDefinitionProvider": true,
        "referencesProvider": true,
        "documentHighlightProvider": true,
        "documentSymbolProvider": true,
        "workspaceSymbolProvider": true,
        "codeActionProvider": false,
        "documentFormattingProvider": true,
        "documentRangeFormattingProvider": true,
        "documentOnTypeFormattingProvider": {
          "firstTriggerCharacter": ";",
          "moreTriggerCharacter": [
            "}"
          ]
        },
        "renameProvider": true,
        "typeCoverageProvider": true,
        "rageProvider": true
      }
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 2,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 6,
            "character": 9
          },
          "end": {
            "line": 6,
            "character": 21
          }
        },
        "title": "b_definition"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 3,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 11,
            "character": 18
          },
          "end": {
            "line": 11,
            "character": 29
          }
        },
        "title": "BB::__construct"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 4,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 14,
            "character": 6
          },
          "end": {
            "line": 14,
            "character": 8
          }
        },
        "title": "CC"
      },
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 11,
            "character": 18
          },
          "end": {
            "line": 11,
            "character": 29
          }
        },
        "title": "BB::__construct"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 5,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 17,
            "character": 6
          },
          "end": {
            "line": 17,
            "character": 8
          }
        },
        "title": "DD"
      },
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 11,
            "character": 18
          },
          "end": {
            "line": 11,
            "character": 29
          }
        },
        "title": "BB::__construct"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 6,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 21,
            "character": 18
          },
          "end": {
            "line": 21,
            "character": 29
          }
        },
        "title": "EE::__construct"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 7,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 26,
            "character": 6
          },
          "end": {
            "line": 26,
            "character": 8
          }
        },
        "title": "FF"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 8,
    "result": [
      {
        "uri": "file://${root_path}/definition.php",
        "range": {
          "start": {
            "line": 40,
            "character": 6
          },
          "end": {
            "line": 40,
            "character": 15
          }
        },
        "title": "HasString"
      }
    ]
  },
  {
    "jsonrpc": "2.0",
    "id": 99,
    "result": null
  }
]
